clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java

```diff
@@ -545,6 +545,15 @@ public ConsumerGroupHeartbeatRequestData buildRequestData() {
             // MemberEpoch - always sent
             data.setMemberEpoch(membershipManager.memberEpoch());
 
+            // Sent all fields if the member is joining/rejoining the group
+            if (membershipManager.state() == MemberState.JOINING) {
+                data.setInstanceId(membershipManager.groupInstanceId().orElse(null));
+                data.setRebalanceTimeoutMs(rebalanceTimeoutMs);
+                data.setSubscribedTopicNames(new ArrayList<>(subscriptions.subscription()));
+                data.setServerAssignor(membershipManager.serverAssignor().orElse(null));
+                data.setTopicPartitions(buildTopicPartitionsList(membershipManager.currentAssignment()));
```

Comment:
The current assignment should be empty when joining, so we don't technically need this. But we need to update `sentFields`! This should be fixed if you apply my suggetion from above